import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java

```diff
@@ -72,12 +81,42 @@ protected void store(
           DicomStreamUtil.dicomStreamWithFileMetaHeader(
               sopInstanceUID, sopClassUID, transferSyntax, countingStream);
 
-      dicomWebClient.stowRs(path, inBuffer);
+      if (redactor != null) {
+        PipedOutputStream pipedOut = new PipedOutputStream();
+        PipedInputStream pipedIn = new PipedInputStream(pipedOut);
 
-      log.info("Received C-STORE for association {}, SOP class {}, TS {}, remote AE {}",
-          association.toString(), sopClassUID, transferSyntax, remoteAeTitle);
-      response.setInt(Tag.Status, VR.US, Status.Success);
+        CompletableFuture<Throwable> futureThrowable = new CompletableFuture<>();
+        association.getApplicationEntity().getDevice().execute(() -> {
+          try {
+            dicomWebClient.stowRs(path, pipedIn);
+            futureThrowable.complete(null);
+          } catch (Throwable e) {
+            futureThrowable.complete(e);
+            try {
+              pipedIn.close();
+            } catch (IOException e1) {
+              log.error("Failed to close pipedIn", e);
+            }
+          }
+        });
+
+        try {
+          redactor.redact(inBuffer, pipedOut);
+        } catch (Exception e){
+          // not sure if this isn't race condition
+          if(!futureThrowable.isDone()) {
+            throw e;
+          }
```

Comment:
If futureThrowable is done but the result is null, then we're not throwing an error here when we should be. Can we just throw e here, and if futureThrowable has an exception we will throw it in the next block?

Reply:
Problem is, exception in redact causes exception in stowRs and I need vice-versa to interrupt redact of failed upload too(so I close pipedIn) . Intention here is to throw up only original cause and not the consequence.

Reply:
If an exception in redact causes a stow exception, then potentially futureThrowable.isDone will already be true here and we will throw the stow-RS exception incorrectly, right?

Reply:
Yes, but I don't see how I could bypass the potential problem. If I added synchronized blocks, it would become matter of which is entered first, for same result.

I could change redact signature to take completable future argument, to be able to complete it inside, thus guaranteeing order. But this would put stuff that is rather irrelevant to redactor library function into it.

Reply:
I'm not sure if a redact exception would cause a stow-RS exception, stow-RS might just hang waiting for input? Is it possible to test that case?

I'm also not sure if pipedIn.close() will necessarily fail the stow-RS call. What if it closes the file at a point where it is almost fully read, so the server can parse it, and accepts it?

An alternative approach I found was to use two threads and an ExecutorCompletionService to get the first result. See an example here: https://stackoverflow.com/a/29116763  it might make this error handling simple if the above is an issue. What do you think?

Reply:
As far as I've tested, both interrupted each-other reliably. I'm not sure how server would handle the prematurely terminated data stream though.

Seems like this approach should work, thanks for finding it. I'd technically use 3 threads - original to pipe PDVInputStream (which can't be read on any other, dcm4che limitation), 2nd to redact and 3rd to send.

Reply:
Yeah, I'm not sure if the executor service strongly guarantees order, but it should make killing the other task easy. If it turns out that we sometimes get the wrong exception, I guess it's OK as long as it's rare.

When we kill the stow thread, I wonder if the server sees this as a cancellation (similar to a CTRL+C in curl) and returns an error. Were you able to see the HTTP response in this error scenario during testing?

Do we need any clean up logic here to avoid leaking threads?

Reply:
After checking for leaks with VisualVM - yes, you are right. Any situation where either thread failed leaked the other, fixed by enclosing all pipe usages in try-with-resources.

If redactor fails while reading or actually redacting, stowRs thread gets 409 with the following details:
```
{"00081198":{"Value":[{"00091097":{"Value":["empty DICOM instance found"],"vr":"LT"},"00090010":{"Value":["GOOGLE"],"vr":"LO"},"00081197":{"Value":[272],"vr":"US"}}],"vr":"SQ"}}
```

I suspect if it somehow failed mid-write, that could be a problem. But doesn't seem likely, at least with current read-all, process-all, write-all sequential implementation.

Reply:
Thanks! So even if the stow-RS command fails at the end of the pipeline, the earlier threads will be interrupted still because they will be writing to a closed stream?

Reply:
Yes, though I suppose redactor will do some unnecessary work in between (since it still reads input, processes it and only then tries to write and fails).

Reply:
Yeah I think that's fine.